### ISSUE TITLE
Determine the best default patterns for core Path auto

### ISSUE THREAD

**Issue by quicksketch** [backdrop/backdrop-issues#940]
When Pathauto was added to core, the default patterns were adjusted when compared to Drupal 7:

Pattern | Drupal 7 | Backdrop
------- | -------- | --------
Node (default) | `content/[node:title]` | `[node:content-type]/[node:title]`
Node Article / Post | - | `posts/[node:title]`
Node Page | - | `[node:title]`
Taxonomy terms | `[term:vocabulary]/[term:name]` | `[term:vocabulary]/[term:name]`
User accounts | `users/[user:name]` | `accounts/[user:name]`

The discussion for these changes was not in the issue queue. Let's review all the reasons for the change and see if these were warranted.

**Note that sites upgrading from Backdrop 1.0 or Drupal 7 have their existing patterns maintained.** This discussion is only about what *default* patterns we should include on new installations.

---
~~PR from @docwilmot at https://github.com/backdrop/backdrop/pull/1555~~
PR from @jenlampton at https://github.com/backdrop/backdrop/pull/1538

**Comment by quicksketch** [comment]
@jenlampton could you chime in here?

Justifications I recall from our discussions include:

- Nixing the default site pattern because it makes creating new patterns in the future more difficult. Currently we have bulk deletion, but only to delete all aliases for all nodes. By removing the default pattern, it makes it so that when new patterns are created, unaliased nodes can have their patterns bulk generated.
- Node Article was left without any pattern, as an obvious one could not be readily determined.
- Node Page was given `[node:title]` as this matches the most common uses of page. e.g. for "About Us", "Location", "Biography", etc. pages usually are not in a particular section or structure.
- Taxonomy terms left alone, as they already have contextual value and make sense most of the time.
- User accounts changed from `users/[user:name]` to `accounts/[user:name]` to match our attempts at consistently referring to users accounts as "accounts" instead of "people" (which is inaccurate) or "users" (which has negative connotations).

**Comment by klonos** [comment]
Fair enough about the page being ```[node:title]```. It seems very sensible to me.

How about ```[node:content-type]/[node:title]``` for the rest of the nodes? It accounts for any future custom content types I believe (for example ```product/milk``` or ```service/cleaning``` etc). What do you guys think?

**Comment by quicksketch** [comment]
My trouble with [node:content-type] is that it is always singular. This works for some content types (e.g. blog) but in my experience most sites tend to use plural forms in their URLs, e.g. photos, videos, etc. These sites usually also make a matching landing page for each section and can sometimes base breadcrumbs off the path in the URL.

**Comment by docwilmot** [comment]
I'm finding this issue strangely difficult to settle in my mind. 

I feel its a bit wrong to have pathauto in core yet leave nodes unaliased; seems a waste. And I grasp the concern about making creating new patterns in the future more difficult but it isnt really *that* difficult.

So far @klonos' suggestion is winning: I'd say `article/this-title` or `product/milk` sounds fine; the users wanting plurals can change it easily enough.

**Comment by klonos** [comment]
> I feel its a bit wrong to have pathauto in core yet leave nodes unaliased; seems a waste. ...

My though precisely. Besides, another "plus" is that ```product/milk``` for example is way better out-of-the-box SEO than ```/node/23```.

**Comment by klonos** [comment]
> My trouble with ```[node:content-type]``` is that it is always singular.

#1145 :wink:

**Comment by sutibun** [comment]
> … but in my experience most sites tend to use plural forms in their URLs, e.g. photos, videos, etc.

My experience probably not as extensive as yours but I've seen the opposite where it's singular. If I remember correctly, even Tumblr has the singular form on their URL `tumblr.com/post/10283939/blah-blah-blah`

For me at least, it seems more "rational" to go with the singular form because a post's URL is pointing to a *single* post. example.com/product/milk

The plural form seems to make more "sense" for the page listing all the posts of that content type example.com/products

Anyway, I also vote for [node:content-type]/[node:title]

> I feel its a bit wrong to have pathauto in core yet leave nodes unaliased; seems a waste. And I grasp the concern about making creating new patterns in the future more difficult but it isnt really *that* difficult.

+1

**Comment by quicksketch** [comment]
I'm coming around to using the singular form as well. The difficulties of making a plural form of content types is fraught with issues, see https://github.com/backdrop/backdrop-issues/issues/1145.

Now that our default content type is called "post" rather than "article", I also like the default out-of-box pattern a bit more.

**Comment by klonos** [comment]
Yeah, it looks like it may take a while to solve #1145 properly and we shouldn't hold this issue on that one when we can ship with sensible defaults out of the box (```[node:content-type]/[node:title]```). Once we sort the pluralization thing, we can implement a token and revise the default pattern to use that instead (```[node:content-type:plural]/[node:title]```).

**Comment by klonos** [comment]
I would like to revive this issue. Lets please eliminate the "node" thing from as many user-facing places as we can (including the site URLs).

Do we have consensus on using `[node:content-type]/[node:title]` for all content as the default/fallback?

**Comment by klonos** [comment]
...can we get this in with 1.4.4 or should we leave it for 1.5.0? Setting 1.x as a milestone for now so that it doesn't fall off the radar.

**Comment by jenlampton** [comment]
@klonos though I like the idea of removing `node/X` from as many places as possible, I don't think we should add any bad defaults until we give people a tool that allows an easy way to change them.

Right now pathauto only lets you delete and regenerate aliases for ALL your nodes. You can't, for example, delete the bad aliases for just your posts, and regenerate only those. You can regenerate aliases via the bulk operation on admin/content, but only 50 nodes at a time. If you have a blog with a gazillion nodes, that gets really frustrating.

What we *do* have right now is the ability to easily generate aliases for all content that is not YET aliased. This means it is easy for people to set aliases for content after-the-fact and fix up all those ugly `node/X`s.

I disagree with Drupal's bad default value of `content/[node:title]` because it adds a useless word `content` that means nothing to people, means nothing to search engines, and decreases the "value" of your page by making the URL longer, and adding a slash, effectively moving the page "farther" from the home page of your site. Add that to the "painful-to-undo" state of per-content-type aliasing, and you have for grumpy experiences.

The only problem I have with `[node:content-type]/[node:title]` is that it's not usually what people want. More often (80%+) they want the plural of the content-type, `products/milk` and `articles/post-title`. People like that plural because if you follow the url backwards to `/articles` or `/products` you also usually see (80%+) a list of articles or products.

As mentioned above, this plural is something we can't programmatically determine, and it's something we don't yet collect when we create the node type (and I'm not sure we should, since it's only useful purpose would be for this alias, and potentially something like simpleviews in core).

> ...can we get this in with 1.4.4

No, let's not rush this. I think if we are going to go with the singular form (something 80%+ people do NOT want) we need to put in a tool that allows it to be easily changed at the same time.

I would like to focus on this for the 1.5 release, however, because I do think getting this right (or at least, better) will improve the editorial experience.

**Comment by docwilmot** [comment]
How about something like this for the Bulk Generate UI:

Generate URL aliases for the following types of un-aliased paths:
- [ ] Content URL aliases                       [ ] delete existing
  - [ ] Page                       [ ] delete existing
  - [ ] Post                       [ ] delete existing
- [ ] Taxonomy term URL aliases                       [ ] delete existing
  - [ ] Tags                       [ ] delete existing
- [ ] User account URL aliases                       [ ] delete existing

**Comment by jenlampton** [comment]
👍 love it!

**Comment by docwilmot** [comment]
@jenlampton tried a PR for this, but the sandbox isnt working :cry:

**Comment by docwilmot** [comment]
If https://github.com/backdrop/backdrop-issues/issues/2077 gets in this will need a PR to get in too! Decisions?! 
- global `[node:content-type]/[node:title]`
- post `posts/[node:title]`

Is this it? I'd say if we're going with `posts/[node:title]` (plural posts) we should ship with a view with path `posts` to avoid confusion.

**Comment by jenlampton** [comment]
Looks good to me. Please open a new issue for the posts view.

On Sep 10, 2016 5:58 AM, "docwilmot" <notifications@github.com> wrote:

If #2077 <https://github.com/backdrop/backdrop-issues/issues/2077> gets in
this will need a PR to get in too! Decisions?!

   - global [node:content-type]/[node:title]
   - post posts/[node:title]

Is this it? I'd say if we're going with posts/[node:title] (plural posts)
we should ship with a view with path posts to avoid confusion.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/backdrop/backdrop-issues/issues/940#issuecomment-246110343>,
or mute the thread

**Comment by quicksketch** [comment]
> this will need a PR to get in too! Decisions?!
> - global [node:content-type]/[node:title]
> - post posts/[node:title]
> Is this it?

I think this is suitable and it's what I would do on my own site.

> I'd say if we're going with posts/[node:title] (plural posts) we should ship with a view with path posts to avoid confusion.

That's a good idea and furthers the concept of having an out-of-box blog, but it should be a separate issue.

**Comment by jenlampton** [comment]
Just getting a PR ready for this one: https://github.com/backdrop/backdrop/pull/1538

BLOCKED BY https://github.com/backdrop/backdrop-issues/issues/2077

**Comment by quicksketch** [comment]
#2077 is in. We are unblocked. Oh yeah! 

![kool1](https://cloud.githubusercontent.com/assets/100206/18427610/3d42d5f8-787e-11e6-88dc-83215ddca973.jpg)

**Comment by quicksketch** [comment]
Hm, PR at https://github.com/backdrop/backdrop/pull/1538, weren't we going to set global to `[node:content-type]/[node:title]`? I only see the `posts/[node:title]` change. Tests are failing, looks like some of them are actually testing the URL in address bar rather than using the internal path at `$_GET['q']`.

**Comment by jenlampton** [comment]
>  weren't we going to set global to [node:content-type]/[node:title]?

We're we? I re-read the recent comments and only saw the posts/[node:title] suggestion. But now that #2077 is in I'm fine with that too.

**Comment by jenlampton** [comment]
Pushed a new PR. testing now. https://github.com/backdrop/backdrop/pull/1538

**Comment by quicksketch** [comment]
Looks good from a config-changes stand-point, but tests are still failing. Once tests are updated the basics of this all look good.

**Comment by jenlampton** [comment]
2:30am...  too sleepy to fix texts... tomorrow!

**Comment by jenlampton** [comment]
Hm, I just spent the past 2 hours doing the same thing it looks like @docwilmot had already done over in https://github.com/docwilmot/backdrop/tree/default-patterns :( whoopsie. Was there a new PR linked here and I missed it?

**Comment by jenlampton** [comment]
Okay, I grabbed @docwilmot's commits and pulled them over into my PR here: 940-default-path-patterns.  Retesting!.

**Comment by docwilmot** [comment]
> Hm, I just spent the past 2 hours doing the same thing it looks like @docwilmot had already done 

Sorry, wasnt meant to be a PR yet, just a push for Travis to report. Glad you found it, sorry about the duplication.

**Comment by jenlampton** [comment]
> sorry about the duplication.

No prob, you'd done more than I had, so still a win to find it!

**Comment by quicksketch** [comment]
Great job! I merged https://github.com/backdrop/backdrop/pull/1538 into 1.x for 1.5.0.
